src/serde/deserialize_utils.rs

```diff
@@ -0,0 +1,359 @@
+        1 => parse_reference_no_offsets(splitted),
+        2 => parse_reference_with_one_offset(splitted),
+        3 => parse_reference_with_two_offsets(splitted),
+        _ => Err(ReferenceParseError::InvalidStringError(String::from(value))),
+    }
+}
+
+fn parse_reference_no_offsets(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split(',').collect::<Vec<_>>()[0] {
+        "cast(ap" => Some(Register::AP),
+        "cast(fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1: 0,
+        offset2: 0,
+        immediate: None,
+        dereference: false,
+    })
+}
+
+fn parse_reference_with_one_offset(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split('(').collect::<Vec<_>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    let mut offset1_str = splitted_value_str[1].split(',').collect::<Vec<_>>()[0].to_string();
+    offset1_str.retain(|c| !r#"()"#.contains(c));
+
+    let offset1: i32 = match offset1_str.parse() {
+        Ok(offset1) => offset1,
+        // for the moment, references with values that overflow i32 are not important, they are just dummy references.
+        Err(e) => match e.kind() {
+            IntErrorKind::PosOverflow => 0,
+            _ => return Err(ReferenceParseError::IntError(e)),
+        },
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1,
+        offset2: 0,
+        immediate: None,
+        dereference: false,
+    })
+}
+
+fn parse_reference_with_two_offsets(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split('[').collect::<Vec<_>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    let mut offset1_str = splitted_value_str[1].to_string();
+    offset1_str.retain(|c| !r#"()]"#.contains(c));
+
+    let offset1: i32 = match offset1_str.parse() {
+        Ok(offset1) => offset1,
+        // for the moment, references with values that overflow i32 are not important, they are just dummy references.
+        Err(e) => match e.kind() {
+            IntErrorKind::PosOverflow => 0,
+            _ => return Err(ReferenceParseError::IntError(e)),
+        },
+    };
+
+    let mut immediate_str = splitted_value_str[2].split(',').collect::<Vec<_>>()[0].to_string();
+    immediate_str.retain(|c| !r#"()"#.contains(c));
+
+    let immediate: BigInt = match immediate_str.parse() {
+        Ok(immediate) => immediate,
+        Err(e) => return Err(ReferenceParseError::BigIntError(e)),
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1,
+        offset2: 0,
+        immediate: Some(immediate),
+        dereference: false,
+    })
+}
```

Comment:
This looks oddly similar to the `dereference` cases. Specially the no offset and one offset cases.

Reply:
Yes you are right, I made a refactor to reduce redundant code